File: cell.rb
require_relative 'distances'

class Cell
    attr_reader :row, :col
    attr_accessor :north, :south, :east, :west

    def initialize(row, col)
      @row, @col = row, col
      @links = {}
    end
    def to_s
      "<Cell object at (#{@row}, #{@col})>"
    end


    def link(other, bidi=true)
      @links[other] = true
      other.link(self, false) if bidi
      self
    end
    def unlink(other, bidi=true)
      @links.delete[other]
      other.unlink(self, false) if bidi
      self
    end
    def links
      @links.keys
    end
    def linked?(other)
      @links.key?(other)
    end

    def neighbors
      list = []
      list << north if north
      list << south if south
      list << east  if east
      list << west  if west
      return list
    end


    def distances
      dstncs = Distances.new(self)
      bndry = [self]

      while bndry.any?
        new_bndry = []

        bndry.each do |cell_|
          cell_.links.each do |linked|
            next if dstncs[linked]
            dstncs[linked] = dstncs[cell_] + 1
            new_bndry << linked
          end
        end

        bndry = new_bndry
      end
      return dstncs
    end

end

File: python/mazes/cell.py

from typing import List

class Cell: 
    directions = {  # IJ-directions, not XY
        'south': (1, 0), # (0,-1)
        'east' : (0, 1), # (1, 0)
        'north': (-1,0), # (0, 1)
        'west' : (0,-1)} # (-1,0)
    
    def __init__(self, row:int, col:int): 
        self.row = row
        self.col = col
        self.links = {}
        for dir in self.directions: 
            setattr(self, dir, None)

    def link(self, other:'Cell', bidir:bool=True): 
        self.links[other] = True
        if bidir: 
            other.link(self, bidir=False)
        return

    def unlink(self, other:'Cell', bidir:bool=True): 
        self.links.pop(other)
        if bidir: 
            other.unlink(self, bidir=False)
        return

    def links(self): 
        return self.links.keys()

    def is_linked(self, other:'Cell') -> bool: 
        return (other in self.links)

    def neighbors(self) -> List['Cell']: 
        them = [link for dir in self.directions 
            if (link := getattr(self, dir))]
        return them


